core/src/apps/bitcoin/sign_tx/helpers.py

```diff
@@ -141,74 +148,93 @@ def confirm_nondefault_locktime(lock_time: int, lock_time_disabled: bool) -> Awa
     return (yield UiConfirmNonDefaultLocktime(lock_time, lock_time_disabled))
 
 
-def request_tx_meta(tx_req: TxRequest, coin: CoinInfo, tx_hash: bytes = None) -> Awaitable[Any]:  # type: ignore
+def request_tx_meta(tx_req: TxRequest, coin: CoinInfo, tx_hash: bytes = None) -> Awaitable[TxAckPrevTxType]:  # type: ignore
+    assert tx_req.details is not None
     tx_req.request_type = TXMETA
     tx_req.details.tx_hash = tx_hash
-    ack = yield tx_req
+    ack = yield TxAckPrevMeta, tx_req
     _clear_tx_request(tx_req)
     return sanitize_tx_meta(ack.tx, coin)
 
 
 def request_tx_extra_data(  # type: ignore
     tx_req: TxRequest, offset: int, size: int, tx_hash: bytes = None
-) -> Awaitable[Any]:
+) -> Awaitable[bytearray]:
+    assert tx_req.details is not None
     tx_req.request_type = TXEXTRADATA
     tx_req.details.extra_data_offset = offset
     tx_req.details.extra_data_len = size
     tx_req.details.tx_hash = tx_hash
-    ack = yield tx_req
+    ack = yield TxAckPrevExtraData, tx_req
     _clear_tx_request(tx_req)
-    return ack.tx.extra_data
+    return ack.tx.extra_data_chunk
 
 
-def request_tx_input(tx_req: TxRequest, i: int, coin: CoinInfo, tx_hash: bytes = None) -> Awaitable[Any]:  # type: ignore
+def request_tx_input(tx_req: TxRequest, i: int, coin: CoinInfo) -> Awaitable[TxAckInputType]:  # type: ignore
+    assert tx_req.details is not None
+    tx_req.request_type = TXINPUT
+    tx_req.details.request_index = i
+    ack = yield TxAckInput, tx_req
+    _clear_tx_request(tx_req)
+    return sanitize_tx_input(ack.tx.input, coin)
+
+
+def request_tx_prev_input(tx_req: TxRequest, i: int, coin: CoinInfo, tx_hash: bytes = None) -> Awaitable[TxAckPrevInputType]:  # type: ignore
+    assert tx_req.details is not None
     tx_req.request_type = TXINPUT
     tx_req.details.request_index = i
     tx_req.details.tx_hash = tx_hash
-    ack = yield tx_req
+    ack = yield TxAckPrevInput, tx_req
+    _clear_tx_request(tx_req)
+    return sanitize_tx_prev_input(ack.tx.input, coin)
+
+
+def request_tx_output(tx_req: TxRequest, i: int, coin: CoinInfo) -> Awaitable[TxAckOutputType]:  # type: ignore  # noqa: F811
+    assert tx_req.details is not None
+    tx_req.request_type = TXOUTPUT
+    tx_req.details.request_index = i
+    ack = yield TxAckOutput, tx_req
     _clear_tx_request(tx_req)
-    return sanitize_tx_input(ack.tx, coin)
+    return sanitize_tx_output(ack.tx.output, coin)
```

Comment:
Definitely thumbs up for separating `request_tx_output()` into two functions.

When handling a replacement transaction, I need to look up the outputs of the original transaction and determine whether they are change-outputs, so I need `address_n` etc. Currently I do it by specifying the `tx_hash` of the original, but that doesn't return the necessary fields. So I imagine that once this is merged, I will add a `tx_hash` field to `request_tx_output()` and use that, right? Then the host will need to be able to somehow figure out whether it's supposed to return `TxAckPrevOutputType` or `TxAckOutputType`. I assume that currently it does that by looking whether `tx_req.details.tx_hash` is set. So we need to give some further indication, like using a different field `tx_req.details.orig_tx_hash` or using some flag, like `tx_req.request_type = TXORIGOUTPUT`. Thoughts?

Reply:
From a very quick look, adding a special `request_type` seems to be the way to go. Otherwise we're asking the host to not only check whether `tx_hash` is set, but also if it has a particular value. Which seems weird.

The other way to do that is to add a field to `TxAckPrevOutputType`? You are sort of looking for a "previous transaction" after all. If `address_n` is the only thing you need to know, that seems like an acceptable thing to do.

Reply:
Both approaches make sense to me, but:

> The other way to do that is to add a field to `TxAckPrevOutputType`? You are sort of looking for a "previous transaction" after all. If `address_n` is the only thing you need to know, that seems like an acceptable thing to do.

We'd need to add all the fields required by `output_is_change()`, so `address_n`, `multisig` and `script_type`.

Reply:
let's discuss this in person